### ISSUE TITLE
Rename _real_XXX() to _next_XXX()?

### ISSUE THREAD

**Issue by rohgarg** [dmtcp/dmtcp#113]
`_real_XXX()` is misleading at times. I think calling it `_next_XXX()` is more accurate. Any thoughts?

**Comment by karya0** [comment]
I definitely like `_next_XXX()` over `_real_XXX()`, but trying to see if there is a better alternative. One way to think about it is that we don't want the plugins to worry about next and previous plugins in the library order. They should only care about their wrapper that then call the "system" function.

**Comment by rohgarg** [comment]
@karya0: 

> One way to think about it is that we don't want the plugins to worry about next and previous plugins in the library order.

I agree.

> They should only care about their wrapper that then call the "system" function.

From what I have seen, when writing a plugin, there are two cases:
 - a plugin may want to delegate the responsibilities of virtualization/ckpt/restore of a resource (which it isn't primarily concerned with) to the _next_ plugin layer, or
 - a plugin may want to handle the virtualization/ckpt/restore of a resource itself.

In the first case, it would make sense to call the `_next_XXX()` function, and in the second case, it would make sense to call the `_real_XXX()` function. This would make the intent explicit, and make the code more readable.

**Comment by gc00** [comment]
I agree that this is a great suggestion.  Eventually, we should also find a way to create either `_libc_XXX()` or maybe even better: ` _external_XXX()`.  The meaning of  `_external_XXX()` is to take the next symbol in library search order that is outside of DMTCP.

The function   `_external_XXX()` could be achieved by recursively using RTLD_NEXT and checking what library we're in, or else by invoking `dl_iterate_phdr()` to walk the list of loaded libraries and trying the equivalent of `dlsym()` on each one whose name doesn't start with `libdmtcp_` or `libdmtcp.so`.  See issue #91 for related implementation details concerning working with libraries loaded by ELF.

**Comment by rohgarg** [comment]
There's a third case, when a plugin may want a call to go through all the definitions from the top in the library search order, i.e., when it calls `XXX()` directly.

I think having these three kinds of calls (or possibly more in the future) would make the plugin API more expressive.
